src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java

```diff
@@ -111,7 +111,8 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
         if (request.isTriggerOnPullRequest()) {
             supportedEvents.add(PULL_REQUEST_DECLINED);
             supportedEvents.add(PULL_REQUEST_DELETED);
-            if (!forbiddenEvents.contains(PULL_REQUEST_FROM_REF_UPDATED) && !request.isMirror()) {
+            if (!forbiddenEvents.contains(PULL_REQUEST_FROM_REF_UPDATED) && !request.isMirror() &&
```

Comment:
I've thought a bit more about this, and while this fixes the immediate problem, I'm not sure we should aim for a minimal fix here. I'm proposing that we change this structure a little in this method. What I'm proposing is
1. we remove the `forbiddenEvents` and instead in the `catch (BitbucketMissingCapabilityException e)` we add the _known_ set of supported events. These cannot change, we know exactly what they are (whereas the return from the server is dynamic and can change over time). This can be a constant so we could just do something like `hooks.addAll(PRE_6_4_WEBHOOK_EVENTS` (or some such)
2. We remove the `hooks.contains(PULL_REQUEST_FROM_REF_UPDATED.getEventId()))` and instead change the `return` to be `return supportedEvents.stream.filter(event -> hooks.contains(event.getEventId())).collect(toSet());`

That way we can always just add events in and the filtering will take care of removing events that are not supported. Maybe we want some logging/better error handling, but we should be able to do that as well should we want to.

What do you both think?

See this as a request for comments/thoughts rather than a request for change.

Reply:
Which events will be set in `PRE_6_4_WEBHOOK_EVENTS`? 
Based on my limited understanding looking at the current code, I feel we don't need `forbiddenEvents` or even `PRE_6_4_WEBHOOK_EVENTS` if we just do this - 

```
            if (!request.isMirror() && hooks.contains(PULL_REQUEST_FROM_REF_UPDATED.getEventId())) {
                supportedEvents.add(PULL_REQUEST_FROM_REF_UPDATED);
            }
```

Reply:
That will break badly under any pre 6.5 BbS version

Reply:
> That will break badly under any pre 6.5 BbS version

I tried the following code in 6.4.0 and all the tests in `TriggerWebhookCreationTest` passed - 

```java
    private Collection<BitbucketWebhookEvent> getEvents(WebhookRegisterRequest request) {
        Set<BitbucketWebhookEvent> supportedEvents = new HashSet<>();
        Set<String> hooks = new HashSet<>();
        try {
            BitbucketWebhookSupportedEvents events = serverCapabilities.getWebhookSupportedEvents();
            hooks.addAll(events.getApplicationWebHooks());
        } catch (BitbucketMissingCapabilityException ignored) {

        }

        if (request.isMirror() && hooks.contains(MIRROR_SYNCHRONIZED.getEventId())) {
            supportedEvents.add(MIRROR_SYNCHRONIZED);
        } else if (request.isTriggerOnRefChange()) {
            supportedEvents.add(REPO_REF_CHANGE);
        }
        if (request.isTriggerOnPullRequest()) {
            supportedEvents.add(PULL_REQUEST_DECLINED);
            supportedEvents.add(PULL_REQUEST_DELETED);
            if (!request.isMirror() && hooks.contains(PULL_REQUEST_FROM_REF_UPDATED.getEventId())) {
                supportedEvents.add(PULL_REQUEST_FROM_REF_UPDATED);
            }
            supportedEvents.add(PULL_REQUEST_MERGED);
            supportedEvents.add(PULL_REQUEST_OPENED);
        }

        return supportedEvents;
    }
```